File: 35.search-insert-position.rs
/*
 * @lc app=leetcode id=35 lang=rust
 *
 * [35] Search Insert Position
 */

// @lc code=start
impl Solution {
    pub fn search_insert(nums: Vec<i32>, target: i32) -> i32 {
        let (mut low, mut high) = (0, nums.len() - 1);

        while low <= high {
            let mid = low + (high - low) / 2;
            if nums[mid] == target{
                return mid as i32;
            } else if target < nums[mid]{
                if mid == 0{
                    return 0;
                }
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        low as i32
    }
}
// @lc code=end

File: 35.search-insert-position.cpp
/*
 * @lc app=leetcode id=35 lang=cpp
 *
 * [35] Search Insert Position
 */

// @lc code=start
class Solution {
 public:
  int searchInsert(vector<int>& nums, int target) {
    int n = nums.size();
    int l = 0, r = n - 1;

    while (l < r) {
      int mid = (r - l) / 2 + l;

      if(nums[mid] == target) return mid;

      if (nums[mid] < target) {
        l = mid + 1;
      } else {
        r = mid;
      }
    }

    return nums[l] < target ? l + 1 : l;
  }
};
// @lc code=end

